TShockAPI/TSPlayer.cs

```diff
@@ -1508,6 +1508,139 @@ public bool GiveItemCheck(int type, string name, int stack, int prefix = 0)
 		/// <param name="stack">The item stack.</param>
 		/// <param name="prefix">The item prefix.</param>
 		public virtual void GiveItem(int type, int stack, int prefix = 0)
+		{
+			if (TShock.Config.Settings.GiveItemsDirectly)
+				GiveItemDirectly(type, stack, prefix);
+			else
+				GiveItemByDrop(type, stack, prefix);
+		}
+
+		private Item EmptySentinelItem = new Item();
+
+		// logic inspired by Terraria.Player.GrabItem
+		private void GiveItemDirectly(int type, int stack, int prefix)
+		{
+			if (ItemID.Sets.IsAPickup[type] || !Main.ServerSideCharacter || this.IsDisabledForSSC)
+			{
+				GiveItemByDrop(type, stack, prefix);
+				return;
+			}
+
+			var item = new Item();
+			item.netDefaults(type);
+			item.stack = stack;
+			item.prefix = (byte)prefix;
+
+			var depleted = (Item item) => (item.type == 0 || item.stack == 0);
+
+			int num = 50;
+			int num2 = 0;
+			if (item.IsACoin)
+			{
+				num2 = -4;
+				num = 54;
+			}
+
+			if (item.FitsAmmoSlot())
+				if (depleted(item = GiveItem_FillAmmo(item)))
+					return;
+
+			for (int i = num2; i < 50; i++)
+			{
+				int num3 = i;
+				if (num3 < 0)
+					num3 = 54 + i;
+
+				if (depleted(item = GiveItem_FillIntoOccupiedSlot(item, num3)))
+					return;
+			}
+
+			if (!item.IsACoin && item.useStyle != 0)
+				for (int i = 0; i < 10; i++)
+					if (depleted(item = GiveItem_FillEmptyInventorySlot(item, i)))
+						return;
+
+			for (int i = num - 1; i >= 0; i--)
+				if (depleted(item = GiveItem_FillEmptyInventorySlot(item, i)))
+					return;
+
+			// oh no, i can't give... guess i gotta spill it on the floor
+			GiveItemByDrop(type, stack, prefix);
```

Comment:
Should be using overflowed `item`? 
```suggestion
			GiveItemByDrop(item.type, item.stack, item.prefix);
```

Reply:
Also maybe could handle going into void vault, though that's just a nice-to-have.